### PR TITLE
Simplify kubernetes_resource_interface

### PR DESCRIPTION
**Pull request by matthchr**
Split validation and defaulting into their own methods added by a different stage.
Conceptually these two interfaces are not directly related to the `KubernetesResource` interface.

### CODE REVIEW

File: v2/tools/generator/internal/codegen/pipeline/apply_kubernetes_resource_interface.go [hunk 1]
```diff
@@ -14,43 +14,36 @@ import (
 	"github.com/Azure/azure-service-operator/v2/tools/generator/internal/interfaces"
 )
 
+const ApplyKubernetesResourceInterfaceStageID = "applyKubernetesResourceInterface"
+
 // ApplyKubernetesResourceInterface ensures that every Resource implements the KubernetesResource interface
 func ApplyKubernetesResourceInterface(idFactory astmodel.IdentifierFactory) Stage {
 	return MakeLegacyStage(
```
Comment (theunrepentantgeek) [inline v2/tools/generator/internal/codegen/pipeline/apply_kubernetes_resource_interface.go:21]: Update this to use `MakeStage()` at the same time - we want to eventually remove `MakeLegacyStage()` completely.

File: v2/tools/generator/internal/codegen/pipeline/apply_kubernetes_resource_interface.go [hunk 2]
```diff
@@ -14,43 +14,36 @@ import (
...
-				} else {
-					result.Add(typeDef)
+				// this is really very ugly; a better way?
+				if _, ok := newResource.SpecType().(astmodel.TypeName); !ok {
+					// the resource Spec was replaced with a new definition; update it
+					// by replacing the named definition:
+					specName := resource.SpecType().(astmodel.TypeName)
+					updatedDefs.Add(astmodel.MakeTypeDefinition(specName, newResource.SpecType()))
+					newResource = newResource.WithSpec(specName)
 				}
```
Comment (theunrepentantgeek) [inline v2/tools/generator/internal/codegen/pipeline/apply_kubernetes_resource_interface.go:41]: Do we need something like a `Redefine(TypeName, Type)` method on `TypeDefinitionSet` that overwrites any existing definition by that name with a new one? 

The resulting code would resemble this:

```
				// this is really very ugly; a better way?
				if newType, ok := newResource.SpecType().(astmodel.TypeName); !ok {
					// the resource Spec was replaced with a new definition; update it
					// by replacing the named definition:
					specName := resource.SpecType().(astmodel.TypeName)
					updatedDefs.Redefine(specName, newType)
					newResource = newResource.WithSpec(specName)
				}
```

**Review by theunrepentantgeek — CHANGES_REQUESTED** [review]

File: v2/tools/generator/internal/codegen/pipeline/apply_kubernetes_resource_interface.go (hunk 1 above)
Comment (matthchr) [reply v2/tools/generator/internal/codegen/pipeline/apply_kubernetes_resource_interface.go:21]: Done

File: v2/tools/generator/internal/codegen/pipeline/apply_kubernetes_resource_interface.go (hunk 2 above)
Comment (matthchr) [reply v2/tools/generator/internal/codegen/pipeline/apply_kubernetes_resource_interface.go:41]: At least in this context though, the `updatedDefs` doesn't have this type defined at all so it's not clear what `Redefine` would do in that case, or how it's different than `Add`?

Comment (theunrepentantgeek) [reply v2/tools/generator/internal/codegen/pipeline/apply_kubernetes_resource_interface.go:41]: I was thinking that maybe a helper method would reduce the number of steps; if we had overloads, I'd suggest `Add(TypeName, Type)` as an overload for `Add(TypeDefinition)`.

**Review by theunrepentantgeek — APPROVED** [review]
Looks good.

**Pull request merged**